// opensea-js/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SHARED_STOREFRONT_LAZY_MINT_ADAPTER_CROSS_CHAIN_ADDRESS = exports.SHARED_STOREFRONT_ADDRESS_GOERLI = exports.SHARED_STOREFRONT_ADDRESS_MAINNET = exports.DEFAULT_ZONE_BY_NETWORK = exports.WETH_ADDRESS_BY_NETWORK = exports.MERKLE_VALIDATOR_MAINNET = exports.API_PATH = exports.API_BASE_TESTNET = exports.API_BASE_MAINNET = exports.MAX_EXPIRATION_MONTHS = exports.INVERSE_BASIS_POINT = void 0;
const types_1 = require("./types");
exports.INVERSE_BASIS_POINT = 10000; // 100 basis points per 1%
exports.MAX_EXPIRATION_MONTHS = 1;
const API_VERSION = 1;
exports.API_BASE_MAINNET = "https://api.opensea.io";
exports.API_BASE_TESTNET = "https://testnets-api.opensea.io";
exports.API_PATH = `/api/v${API_VERSION}`;
exports.MERKLE_VALIDATOR_MAINNET = "0xbaf2127b49fc93cbca6269fade0f7f31df4c88a7";
exports.WETH_ADDRESS_BY_NETWORK = {
    [types_1.Chain.Mainnet]: "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
    [types_1.Chain.Goerli]: "0xb4fbf271143f4fbf7b91a5ded31805e42b2208d6",
};
exports.DEFAULT_ZONE_BY_NETWORK = {
    [types_1.Chain.Mainnet]: "0x0000000000000000000000000000000000000000",
    [types_1.Chain.Goerli]: "0x0000000000000000000000000000000000000000",
};
exports.SHARED_STOREFRONT_ADDRESS_MAINNET = "0x495f947276749ce646f68ac8c248420045cb7b5e";
exports.SHARED_STOREFRONT_ADDRESS_GOERLI = "0x804159144aefb1dc17b171afcefa5b33746c722f";
exports.SHARED_STOREFRONT_LAZY_MINT_ADAPTER_CROSS_CHAIN_ADDRESS = "0xa604060890923ff400e8c6f5290461a83aedacec";
