// opensea-js/lib/__integration_tests__/init.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sdk = exports.walletAddress = exports.LISTING_AMOUNT = exports.TOKEN_ID = exports.TOKEN_ADDRESS = void 0;
var ethers_1 = require("ethers");
var web3_1 = __importDefault(require("web3"));
var types_1 = require("wyvern-js/lib/types");
var constants_1 = require("../__tests__/constants");
var sdk_1 = require("../sdk");
exports.TOKEN_ADDRESS = process.env.SELL_ORDER_CONTRACT_ADDRESS;
exports.TOKEN_ID = process.env.SELL_ORDER_TOKEN_ID;
exports.LISTING_AMOUNT = process.env.LISTING_AMOUNT;
var webProvider = new web3_1.default.providers.HttpProvider("https://eth-mainnet.g.alchemy.com/v2/".concat(constants_1.ALCHEMY_API_KEY));
var rpcProvider = new ethers_1.ethers.providers.JsonRpcProvider("https://eth-mainnet.g.alchemy.com/v2/".concat(constants_1.ALCHEMY_API_KEY));
var wallet = new ethers_1.ethers.Wallet(constants_1.WALLET_PRIV_KEY !== null && constants_1.WALLET_PRIV_KEY !== void 0 ? constants_1.WALLET_PRIV_KEY : "", rpcProvider);
exports.walletAddress = wallet.address;
exports.sdk = new sdk_1.OpenSeaSDK(webProvider, {
    networkName: types_1.Network.Main,
    apiKey: constants_1.MAINNET_API_KEY,
}, function (line) { return console.info("MAINNET: ".concat(line)); }, wallet);
